lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,271 @@
+func (v *VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.Service.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+	defer v.Service.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and its metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
+		op.Errorf("failed to remove volume (%s) on volume store (%s) due to error (%s)", vol.ID, v.Name, err)
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
+
+	//remove volume metadata directory and children
+	if err := target.RemoveAll(v.volMetadataDirPath(vol.ID)); err != nil {
+		op.Errorf("failed to remove metadata for volume (%s) at path (%q) on volume store (%s)", vol.ID, v.volDirPath(vol.ID), v.Name)
+		//FIXME: Should we bail here? the volume is gone at this point...
+	}
+	op.Infof("Successfully removed volume (%s) from volumestore (%s)", vol.ID, v.Name)
+
+	return nil
+}
+
+func (v *VolumeStore) VolumesList(op trace.Operation) ([]*storage.Volume, error) {
+
+	target, err := v.Service.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+	defer v.Service.Unmount(target)
+
+	volFileInfo, err := target.ReadDir(v.volumesDir())
+	if err != nil {
+		return nil, err
+	}
+	var volumes []*storage.Volume
+	var fetchErrors []error
+
+	for _, fileInfo := range volFileInfo {
+
+		if fileInfo.Name() == "." || fileInfo.Name() == ".." {
+			continue
+		}
+
+		volMetadata, err := getMetadata(op, v.volMetadataDirPath(fileInfo.Name()), target)
+		if err != nil {
+			fetchErrors = append(fetchErrors, err)
+			continue
+		}
+
+		volDeviceBacking := NewNFSVolumeDevice(v.Target, v.volDirPath(fileInfo.Name()))
+
+		vol, err := storage.NewVolume(v.SelfLink, fileInfo.Name(), volMetadata, volDeviceBacking)
+		if err != nil {
+			return nil, err
+		}
+
+		volumes = append(volumes, vol)
+	}
+
+	if len(fetchErrors) > 0 {
+		var collectedErrors bytes.Buffer
+		collectedErrors.WriteString("Some Volumes metadata Could not be fetched: \n")
+
+		for index, err := range fetchErrors {
+			collectedErrors.WriteString(fmt.Sprintf("error (%d) : (%s);", index, err))
+		}
+
+		op.Errorf("some volumes were not able to be listed : (%s)", collectedErrors.String())
+		return volumes, errors.New(collectedErrors.String())
+	}
+
+	return volumes, nil
+}
+
+func writeMetadata(op trace.Operation, metadataPath string, info map[string][]byte, target Target) error {
+	//NOTE: right now we do not support updating metadata, thus we make the ID directory here
+	_, err := target.Mkdir(metadataPath, defaultPermissions)
+	if err != nil {
+		return err
+	}
+
+	op.Infof("Writing metadata to (%s)", metadataPath)
+	for fileName, data := range info {
+		targetPath := path.Join(metadataPath, fileName)
+		blobFile, err := target.Create(targetPath, defaultPermissions)
+		defer blobFile.Close()
```

Comment:
You should move this after the `if err != nil` block.  `blobFile` will `nil` in case of error and this will `segfault`.